### PR TITLE
deep sleep api for esp32

### PR DESCRIPTION
**Pull request by liebman**
This is an implementation of deep and light sleep for esp32.  Included is a minimal implementation of RTCPin for RTC GPIO pins for Xtensa chips.

Wakeup sources implemented:
- Timer - wake from sleep via RTC Timer
- Ext0 - wake from sleep via single RTCPin
- Ext1 - wake from sleep via one or more RTCPin

Examples for wake up:
- esp32/examples/sleep_timer.rs
- esp32/examples/sleep_timer_ext0.rs
- esp32/examples/sleep_timer_ext1.rs

Most of the settings for sleep were pulled from `esp-idf` or gleaned by comparing register settings between  `esp-idf-hal` and `esp32-hal` directly after an initial boot and after wakeup from sleep.  Note in particular `RtcSleepConfig::base_settings` which are the `RTC_CTRL` differences after boot between `esp-idf-hal` and `esp32-hal`.

In Deep sleep I'm seeing about 0.15mA with a ESP32-WROOM32 module on a custom board.

## Thank you!

Thank you for your contribution.
Please make sure that your submission includes the following:

### Must

- [x] The code compiles without `errors` or `warnings`.
- [x] All examples work.
- [x] `cargo fmt` was run.
- [x] Your changes were added to the `CHANGELOG.md` in the proper section.
- [x] You updated existing examples or added examples (if applicable).
- [x] Added examples are checked in CI

### Nice to have

- [x] You add a description of your work to this PR.
- [ ] You added proper docs for your newly added features and code.

### CODE REVIEW

**Comment by liebman** [comment]
@jessebraham I'd love thoughts on what the API should be - particularly with respect to ownership, for instance when you're going to wake using ext0 (RTC GPIO pin) the pin needs to be setup to be routed to RTC - and set in the RTC_IO ext_wakeup0 register - should that be handled in sleep or should that be separately managed by the user?

**Comment by bugadani** [comment]
Disclaimer: I'm just shouting in from the sidelines.

That is actually pretty nice, good job!

For what it's worth, I think I would still prefer if creating a wakeup source would immediately configure it. It wouldn't be hard from here - make the user create the WakeupTriggers object, then pass it to each WakeupSource constructor that immediately applies itself (or not if you want them reusable). Then, WakeupTriggers could be used to start the sleep, maybe.

The only difference would be that this way we wouldn't require `dyn` which I believe doesn't optimize pretty well. I'm not sure in this particular case which should be pretty simple but I think LLVM hasn't been particularly skilled in removing vtables at least a few releases ago.

**Comment by liebman** [comment]
I think I see to implement `Drop` for `Ext0WakeupSource` to at least set the pin back to the IO_MUX.  Should i save the values for anything that I change to be restored at `Drop`?

**Comment by liebman** [comment]
> For what it's worth, I think I would still prefer if creating a wakeup source would immediately configure it. It wouldn't be hard from here - make the user create the WakeupTriggers object, then pass it to each WakeupSource constructor that immediately applies itself (or not if you want them reusable). Then, WakeupTriggers could be used to start the sleep, maybe.
> 
> The only difference would be that this way we wouldn't require `dyn` which I believe doesn't optimize pretty well. I'm not sure in this particular case which should be pretty simple but I think LLVM hasn't been particularly skilled in removing vtables at least a few releases ago.

I like this idea, however I have some concerns: (apologies in advance for the rambling below)

There is also a `RtcSleepConfig` instance that would need to be passed in to apply. We could add WakeupTriggers to it. 
 This tracks what parts can be powered down during sleep and what is required to remain powered - and wake sources adjust this on apply. For instance Ext0 keeps the RTC Peripherals powered during sleep.

Also - what  should happen if the a WakeSource instance dropped after it had been applied and before the sleep? We would need to undo the changes to `RtcSleepConfig` as well and if more than one wake source needs a particular setting how do you know when it can be undone.

Maybe make the apply method from the wake source return a `RtcSleepConfig`?  Then the sleep method could take an array of `RtcSleepConfig` and apply them at sleep time.  (still we have the issue of what to do when one was dropped before sleep....

**Comment by liebman** [comment]
well
```
#[derive(Default)]
pub struct SleepConfig<'a, P: RTCPin + Pin> {
    pub timer: Option<TimerWakeupSource>,
    pub ext0: Option<Ext0WakeupSource<'a, P>>,
}
```
has the issue that you have to specify a type for `P` even if `ext0` is being set to `None` - we'd have to type erase it using `dyn`...

**Comment by bugadani** [comment]
You *can* do a trick like I did in the object-chain crate, where you build up a type in compile time. It will not be very nice, though.

**Comment by bjoernQ** [comment]
My two cents: This looks already good and maybe we should just get it in a mergeable shape (checking comments and documentation) and merge it. Then the community can iterate on the API based on real use cases

I don't think having breaking API changes here in the future is going to be a big problem (like we currently have them everywhere)

**Comment by bugadani** [comment]
> (Also I'm sorry it's taken us so long to come to this conclusion!)

Sorry for being a cause of delay here. I have some interest in adapting this PR to the S3, so I'd second getting this merged sooner than later.

**Review by bjoernQ — APPROVED** [review]
LGTM - just want to wait for Jesse's approval before merging

**Review by jessebraham — APPROVED** [review]
Thank you once again for all your work on this, LGTM!

**Comment by liebman** [comment]
> Hi, sorry for commenting on this closed issue. What is the reason for the 100ms delay before going to sleep?

In the examples?  To make sure the print messages are sent before the UART is powered down.

**Comment by liebman** [comment]
> > > Hi, sorry for commenting on this closed issue. What is the reason for the 100ms delay before going to sleep?
> > 
> > 
> > In the examples? To make sure the print messages are sent before the UART is powered down.
> 
> Sorry, I meant here: https://github.com/liebman/esp-hal/blob/a2b0941001a95366759009c3206b980423769e9e/esp-hal-common/src/rtc_cntl/mod.rs#L239

I'd have to assume then that this was left in by mistake.  Please open an issue so we don't loose track of this.

**Pull request merged**
